torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h

```diff
@@ -3,31 +3,162 @@
+      arg_hash ^= ((arg.index << 1) ^ static_cast<size_t>(arg.stype));
+    }
+    size_t output_hash = 0;
+    for (auto output : outputs_) {
+      output_hash ^= ((output.index << 1) ^ static_cast<size_t>(output.stype));
+    }
+    return ((static_cast<size_t>(record_type_) & 0xff) << 56) |
+        ((output_hash & 0xff) << 48) | ((arg_hash & 0xffff) << 32);
+  }
+
+  //! The base virtual equality operator is defined so all child
+  //! classes can utilize the check for the same args and outputs.
+  virtual bool operator==(const RecordFunctor& other) const {
+    auto result = (record_type_ == other.record_type_);
+    result = result && (args_.size() == other.args_.size()) &&
+        (outputs_.size() == other.outputs_.size());
+    if (result) {
+      for (size_t i = 0; i < args_.size(); ++i) {
+        if ((args_[i].index != other.args_[i].index) ||
+            (args_[i].stype != other.args_[i].stype)) {
+          result = false;
+          break;
+        }
+      }
+    }
+    if (result) {
+      for (size_t i = 0; i < outputs_.size(); ++i) {
+        if ((outputs_[i].index != other.outputs_[i].index) ||
+            (outputs_[i].stype != other.outputs_[i].stype)) {
+          result = false;
+          break;
+        }
+      }
+    }
+    return result;
+  }
+
   //! Abstraction for an operation to build this record's nvFuser Fusion IR
   //! piece if the recording has a cache miss.
-  virtual void operator()(FusionDefinition& fd) = 0;
+  virtual void operator()(FusionDefinition& fd) {}
 
+  //! The base print function when printing Record for a given FusionDefinition
+  //! in python formated code.
+  virtual void print(std::ostream& os, bool close_function = true) const {
+    bool first_output = true;
+    for (auto& output : outputs_) {
+      if (first_output) {
+        first_output = false;
+      } else {
+        os << ", ";
+      }
+      if (output.stype == StateType::Scalar) {
+        os << "S";
+      } else if (output.stype == StateType::Tensor) {
+        os << "T";
+      } else {
+        TORCH_INTERNAL_ASSERT(false, "Unsupported StateType");
+      }
+      os << output.index;
+    }
+    if (outputs_.size() > 0) {
+      os << " = "
+         << "fd." << name_ << "(";
+    } else {
+      os << "fd." << name_ << "(";
+    }
+    bool first_arg = true;
+    for (auto& arg : args_) {
+      if (first_arg) {
+        first_arg = false;
+      } else {
+        os << ", ";
+      }
+      if (arg.stype == StateType::Scalar) {
+        os << "S";
+      } else if (arg.stype == StateType::Tensor) {
+        os << "T";
+      } else {
+        TORCH_INTERNAL_ASSERT(false, "Unsupported StateType");
+      }
+      os << arg.index;
+    }
+    if (close_function) {
+      os << ")";
+    }
+  }
+
+  RecordType recordType() const {
+    return record_type_;
+  }
+
+ protected:
   //! Inputs that are indices into the FusionDefinition's Recorded State.
-  std::vector<size_t> args;
+  std::vector<State> args_;
```

Comment:
nitpick: this part looks a little bit strange to me. We are holding a `vector<State>` for args_ & outputs_.
This sorta makes the inheritance from `State` to `Tensor` and `Scalar` meaningless, since we are truncating data here. I don't see we are getting much from that inheritance currently, might want to just remove that and stick with State?

If we want to keep this design to be future proof, then we'd likely want to figure out the ownership and use pointers here instead? I feel a common approach is to take the owner ship to producer and pass naked pointer to consumer. So we'd have something like
```
std::vector<State> args_;
std::vector<std::unique_ptr<State>> outputs_;
```
^^^ I'm not big fan of this neither. If we really wanted polymorphism, I'd suggest we hide that inside `State`.

Reply:
This was changed.  `Tensor` and `Scalar` no longer inherit from `State`.  Just `State` is passed around.